packages/remix-react/single-fetch.ts

```diff
@@ -0,0 +1,203 @@
+  url.searchParams.delete("index");
+  let indexValuesToKeep = [];
+  for (let indexValue of indexValues) {
+    if (indexValue) {
+      indexValuesToKeep.push(indexValue);
+    }
+  }
+  for (let toKeep of indexValuesToKeep) {
+    url.searchParams.append("index", toKeep);
+  }
+
+  return url.href;
+}
+
+// Determine which routes we want to load so we can add a `?_routes` search param
+// for fine-grained revalidation if necessary.  If a route has not yet been loaded
+// via `route.lazy` then we know we want to load it because it's by definition a
+// net-new route.  If it has been loaded then `shouldLoad` will have taken
+// `shouldRevalidate` into consideration.
+//
+// There is a small edge case that _may_ result in a server loader running
+// _somewhat_ unintended, but it's unavoidable:
+// - Assume we have 2 routes, parent and child
+// - Both have `clientLoader`'s and both need to be revalidated
+// - If neither calls `serverLoader`, we won't make the single fetch call
+// - We delay the single fetch call until the **first** one calls `serverLoader`
+// - However, we cannot wait around to know if the other one calls
+//   `serverLoader`, so we include both of them in the `X-Remix-Routes`
+//   header
+// - This means it's technically possible that the second route never calls
+//   `serverLoader` and we never read the response of that route from the
+//   single fetch call, and thus executing that `loader` on the server was
+//   unnecessary.
+function addRevalidationParam(
+  manifest: AssetsManifest,
+  routeModules: RouteModules,
+  matches: DataStrategyMatch[],
+  _url: string
+) {
+  let url = new URL(_url);
+  let genRouteIds = (arr: string[]) =>
+    arr.filter((id) => manifest.routes[id].hasLoader).join(",");
+
+  // By default, we don't include this param and run all matched loaders on the
+  // server.  If _any_ of our matches include a `shouldRevalidate` function _and_
+  // we've determined that the routes we need to load and the matches are
+  // different, then we send the header since they've opted-into fine-grained
+  // caching.  We look at the `routeModules` here instead of the matches since
+  // HDR adds a wrapper for `shouldRevalidate` even if the route didn't have one
+  // initially.
+  // TODO: We probably can get rid of that wrapper once we're strictly on on
+  // single-fetch in v3 and just leverage a needsRevalidation data structure here
+  // to determine what to fetch
+  if (matches.some((m) => routeModules[m.route.id]?.shouldRevalidate)) {
+    let matchedIds = genRouteIds(matches.map((m) => m.route.id));
+    let loadIds = genRouteIds(
+      matches.filter((m) => m.shouldLoad).map((m) => m.route.id)
+    );
+    if (matchedIds !== loadIds) {
+      url.searchParams.set("_routes", loadIds);
+    }
+  }
+
+  return url.href;
+}
+
+function singleFetchUrl(reqUrl: string) {
+  let url = new URL(reqUrl);
+  url.pathname = `${url.pathname === "/" ? "_root" : url.pathname}.data`;
+  return url;
```

Comment:
single fetch URLs are `/path/to/whatever.data`, except for the root route.  To avoid a super odd-looking (and potentially invalid?) URL like `/.data`, we use `/_root.data`

Reply:
`.data`? I imagine it's not `.json` because with RSC the response can be more than just JSON?